metric-anomaly-detector/src/main/java/org/hypertrace/alert/engine/metric/anomaly/detector/evaluator/MetricCache.java

```diff
@@ -0,0 +1,141 @@
+package org.hypertrace.alert.engine.metric.anomaly.detector.evaluator;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import lombok.Getter;
+import lombok.Setter;
+import org.apache.commons.lang3.tuple.Pair;
+import org.hypertrace.alert.engine.eventcondition.config.service.v1.MetricSelection;
+import org.hypertrace.core.query.service.api.ResultSetChunk;
+import org.hypertrace.core.query.service.api.Row;
+import org.hypertrace.core.query.service.api.Value;
+import org.hypertrace.core.query.service.api.ValueType;
+
+class MetricCache {
+
+  private static final int CACHE_EXPIRY_MINUTES = 5;
+
+  private final Cache<MetricSelection, MetricTimeSeries> metricCache;
+  private final QueryRequestHandler queryRequestHandler;
+
+  MetricCache(QueryRequestHandler queryRequestHandler) {
+    this.queryRequestHandler = queryRequestHandler;
+    this.metricCache =
+        CacheBuilder.newBuilder()
+            .expireAfterAccess(Duration.ofMinutes(CACHE_EXPIRY_MINUTES))
+            .build();
+  }
+
+  List<Pair<Long, Double>> getMetricValues(
+      Map<String, String> requestHeaders,
+      MetricSelection metricSelection,
+      String tenantId,
+      long startTimeMillis,
+      long endTimeMillis) {
+    long metricDurationMillis = endTimeMillis - startTimeMillis;
+
+    MetricTimeSeries metricTimeSeries = metricCache.getIfPresent(metricSelection);
+    // no record for this metric selection
+    // if requested time range is earlier than existing
+    // replace data
+    if (null == metricTimeSeries || startTimeMillis < metricTimeSeries.getStartTimeMillis()) {
+      Iterator<ResultSetChunk> iterator =
+          queryRequestHandler.executeQuery(
+              requestHeaders, metricSelection, tenantId, startTimeMillis, endTimeMillis);
+      List<Pair<Long, Double>> dataList = consumeIterator(iterator);
+      metricCache.put(
+          metricSelection,
+          new MetricTimeSeries(startTimeMillis, endTimeMillis, dataList, metricDurationMillis));
+      return dataList;
+    }
+
+    // need to just get diff of the data
+    if (endTimeMillis > metricTimeSeries.getEndTimeMillis()) {
+      Iterator<ResultSetChunk> iterator =
+          queryRequestHandler.executeQuery(
+              requestHeaders,
+              metricSelection,
+              tenantId,
+              metricTimeSeries.getEndTimeMillis(),
+              endTimeMillis);
+      List<Pair<Long, Double>> dataList = consumeIterator(iterator);
+      metricTimeSeries.getDataList().addAll(dataList);
+      metricTimeSeries.setEndTimeMillis(endTimeMillis);
+      metricTimeSeries.setMaxRetentionPeriodMillis(
```

Comment:
I think we want to retain (startTime, endTime) points in TimeSeries. So,  trimOlderData function can take time as an argument and remove the earlier data. This way, we will not need a filter function too.

Reply:
so the thing is that there could me multiple rules with different `ruleDuration` on the same `metricSelection`

Reply:
So, we are retaining the longest window, and so we need filtering. Can you add this in the comment either at trim function or where this variable is defined - basically the usage of this variable?

Reply:
Does this work https://github.com/hypertrace/hypertrace-alert-engine/pull/60/files/2221c0cf837254a36491bef0dc9f5e6fc60b3a27#diff-02446181b5335fb71aa856dfb4f3a2adbcdb28f518092c889f56d5458c31662fR117 ?